genie-web/src/main/java/com/netflix/genie/web/util/StreamBuffer.java

```diff
@@ -0,0 +1,199 @@
+ * @since 4.0.0
+ */
+@ThreadSafe
+@Slf4j
+public class StreamBuffer {
+
+    private final Object lock = new Object();
+    private final AtomicReference<StreamBufferInputStream> inputStreamRef = new AtomicReference<>();
+
+    private boolean closed;
+    private ByteString currentChunk;
+    private int currentChunkWatermark;
+
+    /**
+     * Constructor.
+     */
+    public StreamBuffer() {
+        this.inputStreamRef.set(new StreamBufferInputStream(this));
+    }
+
+    /**
+     * Close this buffer before all data is written due to an error.
+     * Reading will return the end of stream marker after the current chunk (if any) has been consumed.
+     *
+     * @param t the cause for the buffer to be closed.
+     */
+    public void closeForError(final Throwable t) {
+        log.error("Closing buffer due to error: " + t.getClass().getSimpleName() + ": " + t.getMessage());
+        this.closeForCompleted();
+    }
+
+    /**
+     * Close this buffer because all expected data has been written
+     * Reading will return the end of stream marker after all data has been consumed.
+     */
+    public void closeForCompleted() {
+        synchronized (this.lock) {
+            this.closed = true;
+            this.lock.notifyAll();
+        }
+    }
+
+    /**
+     * Append a chunk of data for consumption.
+     * This call may block and not return until some data is read/consumed.
+     *
+     * @param data the data to write into the buffer
+     * @throws IllegalStateException if writing is attempted after the buffer has been closed
+     */
+    public void write(final ByteString data) {
+        synchronized (this.lock) {
+            while (true) {
+                if (this.closed) {
+                    throw new IllegalStateException("Attempting to write after closing");
+                } else if (this.currentChunk == null) {
+                    // Save this chunk so it can be consumed
+                    this.currentChunk = data;
+                    this.currentChunkWatermark = 0;
+                    // Wake up reading thread
+                    this.lock.notifyAll();
+                    return;
+                } else {
+                    // Previous chunk of data is still being consumed. Wait.
+                    try {
+                        this.lock.wait();
+                    } catch (InterruptedException e) {
+                        log.warn("Interrupted while waiting to write next chunk of data");
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Obtain the input stream to read this data.
+     *
+     * @return the input stream
+     * @throws IllegalStateException if invoked multiple times
+     */
+    public InputStream getInputStream() {
+        final InputStream inputStream = this.inputStreamRef.getAndSet(null);
+        if (inputStream == null) {
+            throw new IllegalStateException("Input stream for this buffer is no longer available");
```

Comment:
This is a pretty harsh contract. So for the lifetime of this we can only get one input stream?

Reply:
Yes, it's an ephemeral buffer for in-transit data.
If it gets misused, i'd rather make it very evident and throw as early as possible.